src/_posts/2019-12-30-backstopjs-tutorial.md

```diff
@@ -0,0 +1,257 @@
+---
+author: Mattsi Jansky
+layout: post
+asset-type: post
+title: Testing Your Website for Visual Regressions With BackstopJS
+date: 2019-12-30 00:00:00
+description: This article explains how to set up BackstopJS visual regression tests and some good practices I've picked up from using it in anger
+image:
+    src: /assets/custom/img/blog/2019-12-30-backstopjs/heading.jpg
+    attribution:
+        text: Photo by Alexander Popov on Unsplash
+        href: https://unsplash.com/photos/-1qb8SIBzKY
+
+abstract: How to visual regression test your website with BackstopJS, plus some good practices I've picked up from using it in anger
+tags:
+- testing
+- visual regression testing
+- javascript
+- front-end
+---
+
+## Introduction
+
+Snapshot testing compares a "snapshot" of the output of a prior version of your software to output from the latest version, to check for unintended changes. When a difference is found you either approve it, by updating the expected output snapshot, or fix the cause of the difference.
+
+Visual regression testing is a form of snapshot testing that tests a web front-end. It goes beyond testing the markup or layout by testing the rendered page captured in an emulated browser. As such they can "catch CSS Curve Balls" as BackstopJS says.
+
+BackstopJS is a framework for visual regression testing, written in Javascript. It treats your web service as a black box, so your website doesn't need to be written in Javascript to work with BackstopJS. One of the benefits it offers is a very comprehensive and helpful diff between your snapshots, embedded in a HTML report that it generates. An example below shows how the scrubber diff method allows you to see both test and reference snapshots simultaneously. You can move the red line to change where the boundry lies. There's also the diff method, which highlights anything that's changed in bright pink. Or you can flick between seing the whole reference or test snapshot to spot what's changed.
+
+This article will explain how to set up BackstopJS and some good practices I've picked up from using it in anger. You'll need to have some awareness of Docker Compose and Yarn or NPM.
+
+![An image of BackstopJS's interactive diff between two snapshots of a website. A red line separates each snapshot, demonstrating that the difference between them is how they render an image on the page; one is blurry and the other is not]({{site.baseurl}}/assets/custom/img/blog/2019-12-30-backstopjs/backstopjs-scrubber.png)
+
+## Why
+
+Why use visual regression testing? I'll assume that you appreciate why testing as a general practice is necessary, so here are several scenarios that visual regression tests will catch and other testing techniuqes won't:
```

Comment:
techniques

Reply:
Good catch. Fixed in 41ec5ba1e1223d6aa6aa470b38d2a63fa3715d6f